src/k8s-engine/client.js

```diff
@@ -48,173 +47,321 @@ class ErrorSuppressingStdin extends events.EventEmitter {
+        watch.watch(path, queryParams, wrappedCallback.bind(this, callback), ...extras);
+      },
+    };
+    this.#services = new k8s.ListWatch(
+      '/api/v1/services',
+      wrappedWatch,
+      () => this.#coreV1API.listServiceForAllNamespaces());
+  }
+
+  // Notify that the client the underlying Kubernetes cluster is about to go
+  // away, and we should remove any pending work.
+  destroy() {
+    this.#shutdown = true;
+    for (const [namespace, endpoint, port, server] of this.#servers) {
+      this.#servers.delete(namespace, endpoint, port);
+      server?.close();
     }
+    this.removeAllListeners('service-changed');
+  }
 
-    #_coreV1API = null;
+  /**
+   * @type k8s.CoreV1Api
+   */
+  get #coreV1API() {
+    this.#_coreV1API ||= this.#kubeconfig.makeApiClient(k8s.CoreV1Api);
+    return this.#_coreV1API;
+  }
 
-    /**
-     * Return a pod that is part of a given endpoint and ready to receive traffic.
-     * @param {string} namespace The namespace in which to look for resources.
-     * @param {string} endpointName the name of an endpoint that controls ready pods.
-     * @returns {Promise<k8s.V1Pod?>}
-     */
-    async getActivePod(namespace, endpointName) {
-      console.log(`Attempting to locate ${endpointName} pod...`);
-      // Loop fetching endpoints, until it matches at least one pod.
-      /** @type k8s.V1ObjectReference? */
-      let target = null;
-      // TODO: switch this to using watch.
-      while (!this.#shutdown) {
-        /** @type k8s.V1EndpointsList */
-        const endpoints = await this.#coreV1API.listNamespacedEndpoints(namespace, { headers: { name: endpointName } });
-        target = endpoints?.body?.items?.pop()?.subsets?.pop()?.addresses?.pop()?.targetRef;
-        if (target || this.#shutdown) {
-          break;
-        }
-        console.log(`Could not find ${endpointName} pod (${endpoints ? 'did' : 'did not'} get endpoints), retrying...`);
-        await new Promise(resolve => setTimeout(resolve, 1000));
-      }
-      if (this.#shutdown) {
-        return null;
+  #_coreV1API = null;
+
+  /**
+   * Return a pod that is part of a given endpoint and ready to receive traffic.
+   * @param {string} namespace The namespace in which to look for resources.
+   * @param {string} endpointName the name of an endpoint that controls ready pods.
+   * @returns {Promise<k8s.V1Pod?>}
+   */
+  async #getActivePod(namespace, endpointName) {
+    console.log(`Attempting to locate ${endpointName} pod...`);
+    // Loop fetching endpoints, until it matches at least one pod.
+    /** @type k8s.V1ObjectReference? */
+    let target = null;
+    // TODO: switch this to using watch.
+    while (!this.#shutdown) {
+      /** @type k8s.V1EndpointsList */
+      const endpoints = await this.#coreV1API.listNamespacedEndpoints(namespace, { headers: { name: endpointName } });
+      target = endpoints?.body?.items
+        ?.flatMap(item => item.subsets)?.filter(x => x)
```

Comment:
I'm pretty sure once you get past `items?.` you don't need any more optional chaining operators, as `flatMap` always returns an array.

Reply:
Ah, right, I misunderstood the optional chaining operator; it _short circuits_ on nullish, instead of evaluating to null for the next item in the chain.